bip-0352.mediawiki

```diff
@@ -0,0 +1,471 @@
+|-
+!+24
+|c||e||6||m||u||a||7|| -
+|}
+
+''v31'' (l) is reserved for a backwards incompatible change, if needed. For ''Silent Payments v0'':
+
+* If the receiver's silent payment address version is:
+** ''v0'': check that the data part is exactly 66-bytes. Otherwise, fail
+** ''v1'' through ''v30'': read the first 66-bytes of the data part and discard the remaining bytes (if any)
+** ''v31'': fail
+* Receiver addresses are always [https://github.com/bitcoin/bips/blob/master/bip-0341.mediawiki BIP341] taproot outputs<ref name="why_taproot">'''Why only taproot outputs?''' Providing too much optionality for the protocol makes it difficult to implement and can be at odds with the goal of providing the best privacy. Limiting to taproot outputs helps simplify the implementation significantly while also putting users in the best eventual anonymity set.</ref>
+* The sender should sign with one of the sighash flags ''DEFAULT'', ''ALL'', ''SINGLE'', ''NONE'' (''ANYONECANPAY'' is unsafe). It is strongly recommended implementations use ''SIGHASH_ALL'' (''SIGHASH_DEFAULT'' for taproot inputs) when possible<ref name="why_not_sighash_anyonecanpay">'''Why is it unsafe to use ''SIGHASH_ANYONECANPAY''?''' Since the output address for the receiver is derived from from the sum of the [[#inputs-for-shared-secret-derivation|Inputs For Shared Secret Derivation]] public keys, the inputs must not change once the sender has signed the transaction. If the inputs are allowed to change after the fact, the receiver will not be able to calculate the shared secret needed to find and spend the output. It is currently an open question on how a future version of silent payments could be made to work with new sighash flags such as ''SIGHASH_GROUP'' and ''SIGHASH_ANYPREVOUT''.</ref>
+* Inputs used to derive the shared secret are from the ''[[#inputs-for-shared-secret-derivation|Inputs For Shared Secret Derivation]]'' list
+
+=== Scanning transactions ===
+
+A transaction is a ''Silent Payments v0'' transaction and MUST be scanned if and only if all of the following are true:
+
+* The transaction contains at least one BIP341 taproot output
+* The transaction has at least one input from the ''[[#inputs-for-shared-secret-derivation|Inputs For Shared Secret Derivation]]'' list
+* The transaction does not spend a new, undefined output type (e.g. SegWit versions > 1)<ref name="skip_txs_with_unknown_prevouts">'''Why skip transactions that spend unknown output scripts?''' Skipping transactions that spend unknown output scripts allows us to have a clean upgrade path for Silent Payments by avoiding the need to scan the same transaction multiple times with different rule sets. If a fancy new output type is added in the future and Silent Payments v1 is released with support, we would want to avoid having to first scan the transaction with the silent payment v0 rules and then again with the silent payment v1 rules. Note: this restriction only applies to the inputs of a transaction.</ref>
```

Comment:
Should `e.g. SegWit versions > 1` be `i.e. SegWit versions > 1`? Are there others that matter?

Reply:
CTV was originally proposed as a bare script output (`<32 bytes> OP_CTV`). I believe there was some effort to incorporate it into Tapscript, but the current version of the BIP seems to imply that CTV can be used anywhere, including bare scripts.

Reply:
More importantly, we never specify what the "defined output types" are. I automatically read this as all policy-standard output types, but standardness rules can change. What about larger than x-of-3 bare multisig that are non-standard but consensus-valid? How about bare `OP_1` ([emphemeral anchor](https://bitcoinops.org/en/topics/ephemeral-anchors/))? This shouldn't be left as an exercise for the reader.

Reply:
Great point, I think what we need here is a "Forward compatibility" section. As of today, we have the "Inputs for shared secret derivation" and then anything else can be spent in the transaction, except for Segwit version > 1. We skip these because they might eventually be added to the "Inputs for shared secret derivation" list for a future silent payments version. For everything else, it's fine to spend it in a transaction.

Because of this, we need to specify that if a future soft fork uses an OP_NOP (bare script) for a soft fork, this new output type cannot be added to a future version's "Inputs for shared secret derivation" list, unless using the v31 non-backward compatible upgrade. Additionally, future versions can only add to the "Inputs for shared secret derivation" list unless using the v31 non-backward compatible upgrade.

Reply:
@RubenSomsen same here, might be nice to flesh this out a bit more. IIRC, you also had some concerns with how forward compatibility is currently phrased in the BIP?